From e30a194b08e01bc55a8d093924ddfcdedf08b644 Mon Sep 17 00:00:00 2001
From: Subrata Chatterjee <schatterjee@habana.ai>
Date: Sun, 25 Jul 2021 07:30:05 +0300
Subject: [PATCH] [SW-45031] gaudi2: remove duplicate code

Target is to avoid duplicate code across different
gaudi2 targets(i.e. arm and arc). In this
patch many functions common in gaudi2/arm/common.c
and gaudi2/arc/common.c moved to gaudi2/common/common.c.

Change-Id: Ib3fd215b48261a963225af55cf1807a22162e7a5
Signed-off-by: Subrata Chatterjee <schatterjee@habana.ai>
---

diff --git a/arch/gaudi2/arc/common.c b/arch/gaudi2/arc/common.c
index b773690..66c72bc 100644
--- a/arch/gaudi2/arc/common.c
+++ b/arch/gaudi2/arc/common.c
@@ -8,7 +8,6 @@
 #include <pcie_phy_regs.h>
 #include <psoc_global_conf_regs.h>
 #include <psoc_gpio1_regs.h>
-#include <psoc_qspi_regs.h>
 #include <psoc_efuse_regs.h>
 #include <psoc_arc0_cfg_regs.h>
 #include <psoc_arc0_cfg_masks.h>
@@ -18,11 +17,7 @@
 #include <pcie_dbi_regs.h>
 #include <pcie_dbi_masks.h>
 #include <cpu_ca53_cfg_regs.h>
-#include <cpu_timestamp_regs.h>
 #include <psoc_reset_conf_regs.h>
-#include <cpu_timestamp_regs.h>
-#include <cpu_if_regs.h>
-#include <cpu_if_masks.h>
 
 #include <stdint.h>
 #include <hl_boot_if.h>		/* requires stdint.h */
@@ -42,207 +37,10 @@
 #include <arcboot_export.h>
 #include <pci.h>
 
-#define FW_WORD_NUM		0x2200	/* FW_SIZE / 2 */
 #define PRODUCT_ID_STR_SIZE	0x8
-#define PCIE_LTR_DISABLE		0
-#define PCIE_CAP_HDR_NXT_OFFSET		0x1f8
 
 #define mmPCIE_WRAP_VUART_TX_DATA       0x4C15F60
 #define mmPCIE_WRAP_VUART_TX_STATUS     0x4C15F64
-#define HL_PCIE_LINK_RETRIES	50
-
-u32 get_efuse_read_reg(void)
-{
-	return mmPSOC_EFUSE_READ_CTRL;
-}
-
-bool hl_is_debug_mode(enum hl_debug_hw_func func)
-{
-	u32 boot_strap = regmap_readl(mmPSOC_GLOBAL_CONF_BOOT_STRAP_PINS_L);
-	u32 strap_mask;
-
-	switch (func) {
-	case HL_DEBUG_HW_PCIE:
-		strap_mask = BIT(12); /* Dump Select[4] */
-		break;
-	case HL_DEBUG_HW_DRAM:
-		strap_mask = BIT(11); /* Dump Select[3] */
-		break;
-	case HL_DEBUG_HW_I2C:
-		strap_mask = BIT(10); /* Dump Select[2] */
-		break;
-	default:
-		preboot_puts_hex("Unknown H/W function ", func, false);
-		return 0;
-	}
-
-	return boot_strap & strap_mask;
-}
-
-void preboot_init_pcie_phy_fw_update(void)
-{
-	/* Configure FW update size */
-	regmap_writel(FW_WORD_NUM, mmPCIE_WRAP_PHY_FW_FSM_SIZE);
-
-	/*
-	 * PHY FW FSM EN and Direct FW write mechanism enable
-	 * FW FSM assumes that the data is present in the SRAM
-	 * The FSM will wait for SRAM init done before updating
-	 * the SRAM contents
-	 */
-	regmap_writel(9, mmPCIE_WRAP_PHY_FW_FSM);
-
-	/* Take the PHY out of reset */
-	regmap_writel(1, mmPSOC_GLOBAL_CONF_PCIE_PHY_RST_N);
-}
-
-__weak void preboot_init_pcie_phy(void)
-{
-	/* Reset PCIe PHY */
-	regmap_writel(0, mmPSOC_GLOBAL_CONF_PCIE_PHY_RST_N);
-
-	/*
-	 * Set PCIe PHY CR Para I/F to CFG.
-	 * Select CR PARA I/F instead of JTAG
-	 */
-	regmap_writel(1, mmPCIE_PHY_CR_PARA_SEL_0);
-	regmap_writel(1, mmPCIE_PHY_CR_PARA_SEL_1);
-
-	/* Select indirect access to CR PARA I/F instead of PRBI */
-	regmap_writel(1, mmPCIE_PHY_CR_PARA_CFG);
-
-	preboot_init_pcie_phy_fw_update();
-}
-
-u32 preboot_device_id_get(void)
-{
-	/* TODO: remove hard coding */
-	return PCI_IDS_GAUDI2;
-}
-
-static void preboot_pcie_device_id_vendor_id_set(void)
-{
-	u32 device_id_vendor_id = hl_pci_get_ven_id_dev_id();
-
-	/* Update Vendor ID */
-	regmap_writel(device_id_vendor_id,
-		      mmPCIE_DBI_DEVICE_ID_VENDOR_ID_REG);
-
-	/* Update subsystem Vendor ID */
-	regmap_writel(device_id_vendor_id,
-		      mmPCIE_DBI_SUBSYSTEM_ID_SUBSYSTEM_VENDOR_ID_REG);
-
-}
-
-void preboot_disable_pcie_ltr(void)
-{
-	u32 cap_hdr_cfg, dev_cap2_reg;
-
-	/* PCIe LTR configuration, set bits 31:20 to 0x1f8 */
-	cap_hdr_cfg = regmap_readl(mmPCIE_DBI_MARGIN_EXT_CAP_HDR_REG);
-	cap_hdr_cfg &= ~(PCIE_DBI_MARGIN_EXT_CAP_HDR_REG_NEXT_OFFSET_MASK);
-	cap_hdr_cfg |= (PCIE_CAP_HDR_NXT_OFFSET <<
-		 PCIE_DBI_MARGIN_EXT_CAP_HDR_REG_NEXT_OFFSET_SHIFT);
-	regmap_writel(cap_hdr_cfg, mmPCIE_DBI_MARGIN_EXT_CAP_HDR_REG);
-
-	/* Disable LTR support, set bit-11 to '0' */
-	dev_cap2_reg = regmap_readl(mmPCIE_DBI_DEVICE_CAPABILITIES2_REG);
-	dev_cap2_reg &= ~(PCIE_DBI_DEVICE_CAPABILITIES2_REG_PCIE_CAP_LTR_SUPP_MASK);
-	dev_cap2_reg |= (PCIE_LTR_DISABLE <<
-		 PCIE_DBI_DEVICE_CAPABILITIES2_REG_PCIE_CAP_LTR_SUPP_SHIFT);
-	regmap_writel(dev_cap2_reg, mmPCIE_DBI_DEVICE_CAPABILITIES2_REG);
-}
-
-void preboot_init_pcie(void)
-{
-
-	preboot_init_pcie_phy();
-
-	preboot_pcie_device_id_vendor_id_set();
-
-	/* Disable the EQ REDO */
-	regmap_writel(0x00402801, mmPCIE_DBI_GEN3_RELATED_OFF);
-
-	while (!(regmap_readl(mmPCIE_PHY_PHYX_SRAM_INIT_DONE_0) &&
-		regmap_readl(mmPCIE_PHY_PHYX_SRAM_INIT_DONE_1)))
-		;
-
-	/* Execute the Synopsys W/A */
-	regmap_writel(0xF, mmPCIE_PHY_CR_PARA_CTRL);
-	regmap_writel(0x0004A07D, mmPCIE_PHY_CR_PARA_CMD);
-	udelay(20);
-
-	/* PCIe disable LTR */
-	preboot_disable_pcie_ltr();
-
-	/* Enable and Mask LTSSM (erratum 102) */
-	regmap_writel(3, mmPCIE_AUX_LTSSM_EN);
-}
-
-void preboot_pcie_wait_for_link(void)
-{
-	u8 retry = HL_PCIE_LINK_RETRIES;
-
-	/* init dev id as soon as possible, while waiting for prstn */
-	preboot_device_id_init();
-
-	/* Wait for PRSTn */
-	while (!regmap_readl(mmPSOC_GLOBAL_CONF_PRSTN))
-		;
-
-	/* Only after PRSTn is de-asserted */
-	preboot_pcie_enable();
-
-	preboot_puts("Waiting for PCIe UP status\n");
-
-	while (1) {
-		unsigned int prstn, ltssm, smlh, rdlh;
-
-		prstn = regmap_readl(mmPSOC_GLOBAL_CONF_PRSTN);
-		ltssm = regmap_readl(mmPCIE_AUX_LTSSM_EN);
-		smlh = regmap_readl(mmPCIE_AUX_SMLH_LINK_UP);
-		rdlh = regmap_readl(mmPCIE_AUX_RDLH_LINK_UP);
-
-		if (prstn && smlh && rdlh) {
-			preboot_puts("PCIe link is UP!\n");
-			break;
-		}
-
-		/*
-		 * Do not wait indefinitely on PCIe link up status
-		 * Wait for 5 seconds (50 retries with 100 milli-
-		 * seconds sleep), else bailout.
-		 */
-		retry--;
-		if (!retry) {
-			preboot_puts("Timed-out waiting for PCIe link UP!\n");
-			break;
-		}
-
-		udelay(100000);
-
-		/*
-		 * Some host assert prstn during link up while moving to GEN3
-		 * This cause BAR size to be reset
-		 * We stop link up process by disabling ltssm
-		 * till BAR is reconfigured
-		 */
-		if (!prstn && ltssm) {
-			regmap_writel(0, mmPCIE_AUX_LTSSM_EN);
-			continue;
-		}
-
-		if (prstn && !ltssm) {
-			regmap_writel(3, mmPCIE_AUX_LTSSM_EN);
-			continue;
-		}
-	}
-
-	/* W/A H/W bug that may lead to have a wrong mask (0x1) in LTSSM */
-	regmap_writel(3, mmPCIE_AUX_LTSSM_EN);
-}
-
-void preboot_delay(void) {}
 
 void preboot_putc_virt_uart(const char c)
 {
@@ -253,11 +51,6 @@
 	 */
 	regmap_writel(c, mmPCIE_WRAP_VUART_TX_DATA);
 	regmap_writel(1, mmPCIE_WRAP_VUART_TX_STATUS);
-}
-
-unsigned int get_cs_pad_sel(void)
-{
-	return mmPSOC_GLOBAL_CONF_PAD_SEL_19;
 }
 
 void preboot_sync_btl(void) {}
@@ -305,19 +98,6 @@
 	init_prstn();
 
 	init_arc0();
-}
-
-void spi_ssi_cs0_get(unsigned int *mask, unsigned int *data_offset,
-		     unsigned int *out_en_offset)
-{
-	*mask = BIT(19 % 16);
-	*data_offset = mmPSOC_GPIO1_DATAOUT;
-	*out_en_offset = mmPSOC_GPIO1_OUTENSET;
-}
-
-void spi_base_get(unsigned int *regs)
-{
-	*regs = mmPSOC_QSPI_CTRLR0;
 }
 
 void preboot_copy_image(u64 dst, const void *buf, int len, const char *arch)
@@ -383,17 +163,6 @@
 void ppboot_cp_preboot(u64 loadaddr, u32 offset, u32 size, char *name)
 {
 	/* Not implemented - ARC BootROM copies and validates preboot */
-}
-
-u32 hl_pll_timestamp_base_get(void)
-{
-	return mmCPU_TIMESTAMP_CNTCR;
-}
-
-void pci_dev_specific_hot_rst(void)
-{
-	regmap_writel(0x10, mmPCIE_CORE_HOT_RST);
-	regmap_writel(3, mmPCIE_AUX_LTSSM_EN);
 }
 
 static u32 ppboot_flash_image_offset(u8 image_ptr)
@@ -469,18 +238,4 @@
 {
 	*loadaddr = copy_preboot();
 	return true;
-}
-
-void pci_dev_specific_flr(void)
-{
-	regmap_writel(1, mmPCIE_AUX_PF_FLR_ACTIVE);
-	regmap_writel(1, mmPCIE_AUX_PF_FLR_DONE);
-	regmap_writel(1, mmPCIE_AUX_FLR_INT);
-
-	/*
-	 * Clear PCIe SPI INTR. This is required to clear
-	 * FLR interrupt as recommended by HW team.
-	 */
-	regmap_writel(CPU_IF_PCIE_SPI_INTR_CLR_VAL_MASK,
-			mmCPU_IF_PCIE_SPI_INTR_CLR);
 }
diff --git a/arch/gaudi2/arc/common.h b/arch/gaudi2/arc/common.h
index d6fc9ff..206ce0e 100644
--- a/arch/gaudi2/arc/common.h
+++ b/arch/gaudi2/arc/common.h
@@ -57,4 +57,6 @@
 #define ARC_RST_VALUE		0x110
 #define ARC_RST_VEC_ADDR	0x0
 
+#define FW_WORD_NUM		0x2200	/* FW_SIZE / 2 */
+
 #endif /* COMMON_GAUDI2_ARC_GNU_H_ */
diff --git a/arch/gaudi2/arm/common.c b/arch/gaudi2/arm/common.c
index 76ef307..2e570b8 100644
--- a/arch/gaudi2/arm/common.c
+++ b/arch/gaudi2/arm/common.c
@@ -7,21 +7,13 @@
 
 #include <gaudi2_blocks_linux_driver.h>
 #include <psoc_gpio1_regs.h>
-#include <psoc_qspi_regs.h>
-#include <psoc_efuse_regs.h>
 #include <psoc_global_conf_regs.h>
 #include <psoc_global_conf_masks.h>
 #include <pcie_aux_regs.h>
-#include <pcie_core_regs.h>
 #include <pcie_wrap_regs.h>
-#include <pcie_phy_regs.h>
-#include <pcie_phy_masks.h>
 #include <pcie_dbi_regs.h>
 #include <pcie_dbi_masks.h>
-#include <cpu_if_regs.h>
-#include <cpu_if_masks.h>
 
-#include <cpu_timestamp_regs.h>
 #include <psoc_arc0_cfg_regs.h>
 #include <psoc_arc0_cfg_masks.h>
 #include <psoc_arc1_cfg_regs.h>
@@ -45,9 +37,6 @@
 
 #include "common.h"
 
-#define FW_WORD_NUM		0x2000	/* FW_SIZE / 2 */
-#define PCIE_LTR_DISABLE		0
-#define PCIE_CAP_HDR_NXT_OFFSET		0x1f8
 
 #define ARC0_ICCM_BASE_ADDR		0x1000007FF7000000ull
 #define ARC0_ICCM_SIZE			0x20000			/* 128KB */
@@ -58,31 +47,6 @@
 #define ARC1_DCCM_BASE_ADDR		0x1000007FF7060000ull
 #define ARC1_DCCM_SIZE			0x8000			/* 32KB */
 
-#define DUMP_SEL_SRNS_MASK		0x1
-#define DUMP_SEL_SSC_MASK		0x2
-
-bool hl_is_debug_mode(enum hl_debug_hw_func func)
-{
-	u32 boot_strap = regmap_readl(mmPSOC_GLOBAL_CONF_BOOT_STRAP_PINS_L);
-	u32 strap_mask;
-
-	switch (func) {
-	case HL_DEBUG_HW_PCIE:
-		strap_mask = BIT(12); /* Dump Select[4] */
-		break;
-	case HL_DEBUG_HW_DRAM:
-		strap_mask = BIT(11); /* Dump Select[3] */
-		break;
-	case HL_DEBUG_HW_I2C:
-		strap_mask = BIT(10); /* Dump Select[2] */
-		break;
-	default:
-		preboot_puts_hex("Unknown H/W function ", func, false);
-		return 0;
-	}
-
-	return boot_strap & strap_mask;
-}
 
 u64 lkd_dev_spec_get_img_addr(uintptr_t base)
 {
@@ -90,311 +54,9 @@
 	return ((base ^ SDRAM_BASE) | SDRAM_BASE_FULL);
 }
 
-static void pcie_update_reg_range(u32 addr, u32 addr_end, u32 value)
-{
-	for (; addr <= addr_end ; addr += 4)
-		regmap_writel(value, addr);
-}
-
-static void pcie_wait_for_sram_init_done(void)
-{
-	uint8_t sram_init_done = 0;
-
-	do {
-		sram_init_done =
-			regmap_readl(mmPCIE_PHY_PHYX_SRAM_INIT_DONE_0);
-	} while (sram_init_done != 1);
-
-	sram_init_done = 0;
-	do {
-		sram_init_done =
-			regmap_readl(mmPCIE_PHY_PHYX_SRAM_INIT_DONE_1);
-	} while (sram_init_done != 1);
-
-	sram_init_done = 0;
-	do {
-		sram_init_done =
-			regmap_readl(mmPCIE_PHY_PHYX_SRAM_INIT_DONE_2);
-	} while (sram_init_done != 1);
-
-	sram_init_done = 0;
-	do {
-		sram_init_done =
-			regmap_readl(mmPCIE_PHY_PHYX_SRAM_INIT_DONE_3);
-	} while (sram_init_done != 1);
-}
-
-static void pcie_phy_set_sram_ext_ld_done(void)
-{
-	regmap_writel(1, mmPCIE_PHY_PHY_SRAM_EXT_LD_DONE);
-}
-
-static void pcie_phy_set_mpll_a_b_ssc_enable(void)
-{
-
-	pcie_update_reg_range(mmPCIE_PHY_PHYX_MPLLB_SSC_EN_0,
-		mmPCIE_PHY_PHYX_MPLLB_SSC_EN_3, 0x1);
-	pcie_update_reg_range(mmPCIE_PHY_PHYX_MPLLA_SSC_EN_0,
-		mmPCIE_PHY_PHYX_MPLLA_SSC_EN_3, 0x1);
-}
-
-static void pcie_phy_set_mpll_a_b_mul_bw(uint8_t separate_reference)
-{
-	uint32_t reg;
-
-	if (separate_reference) {
-		reg = 0xA0;
-		regmap_writel(0x00250025,
-			mmPCIE_PHY_PROTOCOL0_EXT_MPLLA_BANDWIDTH);
-		regmap_writel(0x00250025,
-			mmPCIE_PHY_PROTOCOL0_EXT_MPLLB_BANDWIDTH);
-	} else {
-		reg = 0x50;
-		regmap_writel(0x00530053,
-			mmPCIE_PHY_PROTOCOL0_EXT_MPLLA_BANDWIDTH);
-		regmap_writel(0x01C301C3,
-			mmPCIE_PHY_PROTOCOL0_EXT_MPLLB_BANDWIDTH);
-	}
-
-	regmap_writel(reg, mmPCIE_PHY_PROTOCOL0_EXT_MPLLA_MULTIPLIER);
-	regmap_writel(0x7D, mmPCIE_PHY_PROTOCOL0_EXT_MPLLB_MULTIPLIER);
-}
-
-static void pcie_phy_set_ext_tx_rx_ctrl(void) {}
-static void pcie_phy_set_protocol0_ext_tx_rx(void) {}
-
-static uint8_t get_srns_mode_from_strap(uint32_t boot_strap)
-{
-	uint32_t dump_sel = (boot_strap &
-		PSOC_GLOBAL_CONF_BOOT_STRAP_PINS_L_DUMP_SEL_MASK) >>
-		PSOC_GLOBAL_CONF_BOOT_STRAP_PINS_L_DUMP_SEL_SHIFT;
-
-	/* dumpsel[0] is SRNS */
-	return (dump_sel & DUMP_SEL_SRNS_MASK);
-}
-
-static uint8_t get_ssc_en_from_strap(uint32_t boot_strap)
-{
-	uint32_t dump_sel = (boot_strap &
-		PSOC_GLOBAL_CONF_BOOT_STRAP_PINS_L_DUMP_SEL_MASK) >>
-		PSOC_GLOBAL_CONF_BOOT_STRAP_PINS_L_DUMP_SEL_SHIFT;
-
-	/* dumpsel[1] is SSC EN */
-	return !!(dump_sel & DUMP_SEL_SSC_MASK);
-}
-
-uint8_t get_sris_mode_from_strap(void)
-{
-	uint32_t boot_strap;
-
-	boot_strap = regmap_readl(mmPSOC_GLOBAL_CONF_BOOT_STRAP_PINS_H);
-
-	return !!(boot_strap & PSOC_GLOBAL_CONF_BOOT_STRAP_PINS_H_SRIS_MODE_MASK);
-}
-
-static void pcie_wa_snps_case_01038891(void)
-{
-	uint8_t separate_reference = 0;
-	uint8_t ssc_en = 0, srns = 0;
-	uint32_t boot_strap = regmap_readl(mmPSOC_GLOBAL_CONF_BOOT_STRAP_PINS_L);
-
-	separate_reference = get_sris_mode_from_strap();
-	if (separate_reference) {
-		regmap_writel(1, mmPCIE_PHY_PIPE_RX_SRIS_MODE_EN);
-		regmap_writel(1, mmPCIE_AUX_APP_SRIS_MODE);
-	}
-	/* SRNS mode enable */
-	srns = get_srns_mode_from_strap(boot_strap);
-
-	if (!srns)
-		separate_reference = 1;
-
-	/* spread spectum clock enable */
-	ssc_en = get_ssc_en_from_strap(boot_strap);
-
-	/* TODO: Check fuse to see if SNPS WA needs to be enabled  */
-	if (ssc_en) {
-		pcie_phy_set_mpll_a_b_ssc_enable();
-
-		regmap_writel(0x5000, mmPCIE_PHY_EXT_MPLLB_SSC_PEAK);
-		regmap_writel(0x19CE,
-				mmPCIE_PHY_EXT_MPLLB_SSC_STEP_SIZE);
-		regmap_writel(0x6666, mmPCIE_PHY_EXT_MPLLA_SSC_PEAK);
-		regmap_writel(0x2108,
-				mmPCIE_PHY_EXT_MPLLA_SSC_STEP_SIZE);
-	}
-
-	pcie_phy_set_ext_tx_rx_ctrl();
-	pcie_phy_set_mpll_a_b_mul_bw(separate_reference);
-
-	pcie_phy_set_protocol0_ext_tx_rx();
-
-	regmap_writel(1, mmPCIE_PHY_PHY_EXT_CTRL_SEL);
-}
-
-void preboot_init_pcie_phy_fw_update(void)
-{
-	/* Configure FW update size */
-	regmap_writel(FW_WORD_NUM, mmPCIE_WRAP_PHY_FW_FSM_SIZE);
-
-	/*
-	 * PHY FW FSM EN and Direct FW write mechanism enable
-	 * FW FSM assumes that the data is present in the SRAM
-	 * The FSM will wait for SRAM init done before updating
-	 * the SRAM contents
-	 */
-	regmap_writel(9, mmPCIE_WRAP_PHY_FW_FSM);
-
-	/* Take the PHY out of reset */
-	regmap_writel(1, mmPSOC_GLOBAL_CONF_PCIE_PHY_RST_N);
-}
-
-__weak void preboot_init_pcie_phy(void)
-{
-	/* Reset PCIe PHY */
-	regmap_writel(0, mmPSOC_GLOBAL_CONF_PCIE_PHY_RST_N);
-
-	/*
-	 * Set PCIe PHY CR Para I/F to CFG.
-	 * Select CR PARA I/F instead of JTAG
-	 */
-	regmap_writel(1, mmPCIE_PHY_CR_PARA_SEL_0);
-	regmap_writel(1, mmPCIE_PHY_CR_PARA_SEL_1);
-	regmap_writel(1, mmPCIE_PHY_CR_PARA_SEL_2);
-	regmap_writel(1, mmPCIE_PHY_CR_PARA_SEL_3);
-
-	/* Execute synopsys workaround (01038891) */
-	pcie_wa_snps_case_01038891();
-
-	/* Select indirect access to CR PARA I/F instead of PRBI */
-	regmap_writel(1, mmPCIE_PHY_CR_PARA_CFG);
-
-	preboot_init_pcie_phy_fw_update();
-}
-
-void preboot_disable_pcie_ltr(void)
-{
-	u32 cap_hdr_cfg, dev_cap2_reg;
-
-	/* PCIe LTR configuration, set bits 31:20 to 0x1f8 */
-	cap_hdr_cfg = regmap_readl(mmPCIE_DBI_MARGIN_EXT_CAP_HDR_REG);
-	cap_hdr_cfg &= ~(PCIE_DBI_MARGIN_EXT_CAP_HDR_REG_NEXT_OFFSET_MASK);
-	cap_hdr_cfg |= (PCIE_CAP_HDR_NXT_OFFSET <<
-		 PCIE_DBI_MARGIN_EXT_CAP_HDR_REG_NEXT_OFFSET_SHIFT);
-	regmap_writel(cap_hdr_cfg, mmPCIE_DBI_MARGIN_EXT_CAP_HDR_REG);
-
-	/* Disable LTR support, set bit-11 to '0' */
-	dev_cap2_reg = regmap_readl(mmPCIE_DBI_DEVICE_CAPABILITIES2_REG);
-	dev_cap2_reg &= ~(PCIE_DBI_DEVICE_CAPABILITIES2_REG_PCIE_CAP_LTR_SUPP_MASK);
-	dev_cap2_reg |= (PCIE_LTR_DISABLE <<
-		 PCIE_DBI_DEVICE_CAPABILITIES2_REG_PCIE_CAP_LTR_SUPP_SHIFT);
-	regmap_writel(dev_cap2_reg, mmPCIE_DBI_DEVICE_CAPABILITIES2_REG);
-}
-
-u32 preboot_device_id_get(void)
-{
-	return PCI_IDS_GAUDI2;
-}
-
-void preboot_init_pcie(void)
-{
-	preboot_init_pcie_phy();
-	u32 ven_id_dev_id = hl_pci_get_ven_id_dev_id();
-
-	/* Update Vendor ID */
-	regmap_writel(ven_id_dev_id, mmPCIE_DBI_DEVICE_ID_VENDOR_ID_REG);
-
-	/* Update subsystem Vendor ID */
-	regmap_writel(ven_id_dev_id,
-		      mmPCIE_DBI_SUBSYSTEM_ID_SUBSYSTEM_VENDOR_ID_REG);
-
-	/* Disable the EQ REDO */
-	regmap_writel(0x00402801, mmPCIE_DBI_GEN3_RELATED_OFF);
-
-	/* Wait for init_done before executing W/A */
-	pcie_wait_for_sram_init_done();
-
-	pcie_phy_set_sram_ext_ld_done();
-
-	/* Execute the Synopsys W/A */
-	regmap_writel(0xF, mmPCIE_PHY_CR_PARA_CTRL);
-	regmap_writel(0x0004A07D, mmPCIE_PHY_CR_PARA_CMD);
-	udelay(20);
-
-	/* PCIe disable LTR */
-	preboot_disable_pcie_ltr();
-
-	/* Enable and Mask LTSSM (erratum 102) */
-	regmap_writel(3, mmPCIE_AUX_LTSSM_EN);
-}
-
-void preboot_pcie_wait_for_link(void)
-{
-	u8 retry = HL_PCIE_LINK_RETRIES;
-
-	/* init dev id as soon as possible, while waiting for prstn */
-	preboot_device_id_init();
-
-	/* Wait for PRSTn */
-	while (!regmap_readl(mmPSOC_GLOBAL_CONF_PRSTN))
-		;
-
-	/* Only after PRSTn is de-asserted */
-	preboot_pcie_enable();
-
-	preboot_puts("Waiting for PCIe UP status\n");
-
-	while (1) {
-		unsigned int prstn, ltssm, smlh, rdlh;
-
-		prstn = regmap_readl(mmPSOC_GLOBAL_CONF_PRSTN);
-		ltssm = regmap_readl(mmPCIE_AUX_LTSSM_EN);
-		smlh = regmap_readl(mmPCIE_AUX_SMLH_LINK_UP);
-		rdlh = regmap_readl(mmPCIE_AUX_RDLH_LINK_UP);
-
-		if (prstn && smlh && rdlh) {
-			preboot_puts("PCIe link is UP!\n");
-			break;
-		}
-
-		/*
-		 * Do not wait indefinitely on PCIe link up status
-		 * Wait for 5 seconds (50 retries with 100 milli-
-		 * seconds sleep), else bailout.
-		 */
-		retry--;
-		if (!retry) {
-			preboot_puts("Timed-out waiting for PCIe link UP!\n");
-			break;
-		}
-
-		udelay(100000);
-
-		/*
-		 * Some host assert prstn during link up while moving to GEN3
-		 * This cause BAR size to be reset
-		 * We stop link up process by disabling ltssm
-		 * till BAR is reconfigured
-		 */
-		if (!prstn && ltssm) {
-			regmap_writel(0, mmPCIE_AUX_LTSSM_EN);
-			continue;
-		}
-
-		if (prstn && !ltssm) {
-			regmap_writel(3, mmPCIE_AUX_LTSSM_EN);
-			continue;
-		}
-	}
-
-	/* W/A H/W bug that may lead to have a wrong mask (0x1) in LTSSM */
-	regmap_writel(3, mmPCIE_AUX_LTSSM_EN);
-}
-
-void preboot_delay(void) {}
-
 #define mmPCIE_WRAP_VUART_TX_DATA	0x4C14F60
 #define mmPCIE_WRAP_VUART_TX_STATUS	0x4C14F64
+
 void preboot_putc_virt_uart(const char c)
 {
 	/*
@@ -404,11 +66,6 @@
 	 */
 	regmap_writel(c, mmPCIE_WRAP_VUART_TX_DATA);
 	regmap_writel(1, mmPCIE_WRAP_VUART_TX_STATUS);
-}
-
-unsigned int get_cs_pad_sel(void)
-{
-	return mmPSOC_GLOBAL_CONF_PAD_SEL_19;
 }
 
 /*
@@ -494,19 +151,6 @@
 	preboot_puts("5\n");
 }
 
-void spi_ssi_cs0_get(unsigned int *mask, unsigned int *data_offset,
-		     unsigned int *out_en_offset)
-{
-	*mask = BIT(19 % 16);
-	*data_offset = mmPSOC_GPIO1_DATAOUT;
-	*out_en_offset = mmPSOC_GPIO1_OUTENSET;
-}
-
-void spi_base_get(unsigned int *regs)
-{
-	*regs = mmPSOC_QSPI_CTRLR0;
-}
-
 void image_start(u64 loadaddr)
 {
 	void (*image_entry)(void) = (void *)loadaddr;
@@ -564,17 +208,6 @@
 	spi_copy((void *)loadaddr, offset, size, name);
 }
 
-u32 hl_pll_timestamp_base_get(void)
-{
-	return mmCPU_TIMESTAMP_CNTCR;
-}
-
-void pci_dev_specific_hot_rst(void)
-{
-	regmap_writel(0x10, mmPCIE_CORE_HOT_RST);
-	regmap_writel(3, mmPCIE_AUX_LTSSM_EN);
-}
-
 bool preboot_is_valid_address_range(u64 addr, int len)
 {
 	if (addr >= HL_SP_SRAM_BASE &&
@@ -595,26 +228,7 @@
 	return false;
 }
 
-u32 get_efuse_read_reg(void)
-{
-	return mmPSOC_EFUSE_READ_CTRL;
-}
-
 u32 dev_pci_dbi_cfg_offset_get(void)
 {
 	return (u32)mmPCIE_DBI_BASE & 0xFFFFFFF;
-}
-
-void pci_dev_specific_flr(void)
-{
-	regmap_writel(1, mmPCIE_AUX_PF_FLR_ACTIVE);
-	regmap_writel(1, mmPCIE_AUX_PF_FLR_DONE);
-	regmap_writel(1, mmPCIE_AUX_FLR_INT);
-
-	/*
-	 * Clear PCIe SPI INTR. This is required to clear
-	 * FLR interrupt as recommended by HW team.
-	 */
-	regmap_writel(CPU_IF_PCIE_SPI_INTR_CLR_VAL_MASK,
-			mmCPU_IF_PCIE_SPI_INTR_CLR);
 }
diff --git a/arch/gaudi2/arm/common.h b/arch/gaudi2/arm/common.h
index 2a1ccb1..703fe55 100644
--- a/arch/gaudi2/arm/common.h
+++ b/arch/gaudi2/arm/common.h
@@ -18,7 +18,6 @@
 #define SDRAM_BASE_FULL			0x1000FFFFFD000000ull
 #endif
 
-#define HL_PCIE_LINK_RETRIES		50
 #define HL_FW_BOOT_PTRS_BASE		0x80000
 #define HL_SPI_LEGACY_PREBOOT_OFF	0xC0000
 
@@ -49,7 +48,6 @@
 #define ARC0_MEM_SIZE			(ARC0_ICCM_SIZE + ARC0_DCCM_SIZE)
 #define ARC1_MEM_SIZE			(ARC1_ICCM_SIZE + ARC1_DCCM_SIZE)
 
-#define HL_PCIE_LINK_RETRIES		50
 #define HL_INIT_SP_SRAM_RETRIES		50
 
 #define BOOT_INFO_BASE			HL_SP_SRAM_BASE
@@ -106,5 +104,7 @@
 
 #define I2C_SLAVE_DEFAULT_INTR_MASK			0x48FF
 
+#define FW_WORD_NUM		0x2000	/* FW_SIZE / 2 */
+
 #endif /* COMMON_GAUDI2_ARM_H_ */
 
diff --git a/arch/gaudi2/common/common.c b/arch/gaudi2/common/common.c
index c9c2c2d..ae1a7c2 100644
--- a/arch/gaudi2/common/common.c
+++ b/arch/gaudi2/common/common.c
@@ -6,19 +6,43 @@
  */
 
 #include <stdint.h>
-#include <preboot.h>
-#include <cpu_if_regs.h>
-#include <cmd_module.h>
+
 #include <psoc_global_conf_regs.h>
+#include <psoc_global_conf_masks.h>
+#include <psoc_gpio1_regs.h>
+#include <psoc_qspi_regs.h>
+#include <psoc_efuse_regs.h>
+#include <pcie_phy_regs.h>
+#include <pcie_phy_masks.h>
+#include <pcie_wrap_regs.h>
+#include <pcie_dbi_regs.h>
+#include <pcie_dbi_masks.h>
+#include <pcie_aux_regs.h>
+#include <pcie_core_regs.h>
+#include <cpu_timestamp_regs.h>
+#include <cpu_if_regs.h>
+#include <cpu_if_masks.h>
+#include <preboot.h>
+#include <cmd_module.h>
 #include <gaudi2/gaudi2_reg_map.h>
 #include <psoc_reset_conf_regs.h>
 #include <psoc_reset_conf_masks.h>
 #include <gaudi2_blocks_linux_driver.h>
-#include <pll.h>
-#include <uart.h>
-#include "common.h"
 #include "preboot/thermal.h"
 #include <string.h>
+
+#include <common/pci_ids.h>
+
+#include <uart.h>
+#include <pll.h>
+#include <pci.h>
+
+#define PCIE_LTR_DISABLE		0
+#define PCIE_CAP_HDR_NXT_OFFSET		0x1f8
+#define HL_PCIE_LINK_RETRIES             50
+
+#define DUMP_SEL_SRNS_MASK		0x1
+#define DUMP_SEL_SSC_MASK		0x2
 
 void lkd_dev_spec_regs(struct cpu_dyn_regs *regs)
 {
@@ -100,3 +124,385 @@
 	regmap_writel(0, mmPSOC_RESET_CONF_TS_IF_UNIT_RST);
 	regmap_writel(1, mmPSOC_RESET_CONF_UNIT_RST_N);
 }
+
+bool hl_is_debug_mode(enum hl_debug_hw_func func)
+{
+	u32 boot_strap = regmap_readl(mmPSOC_GLOBAL_CONF_BOOT_STRAP_PINS_L);
+	u32 strap_mask;
+
+	switch (func) {
+	case HL_DEBUG_HW_PCIE:
+		strap_mask = BIT(12); /* Dump Select[4] */
+		break;
+	case HL_DEBUG_HW_DRAM:
+		strap_mask = BIT(11); /* Dump Select[3] */
+		break;
+	case HL_DEBUG_HW_I2C:
+		strap_mask = BIT(10); /* Dump Select[2] */
+		break;
+	default:
+		preboot_puts_hex("Unknown H/W function ", func, false);
+		return 0;
+	}
+
+	return boot_strap & strap_mask;
+}
+
+void preboot_init_pcie_phy_fw_update(void)
+{
+	/* Configure FW update size */
+	regmap_writel(FW_WORD_NUM, mmPCIE_WRAP_PHY_FW_FSM_SIZE);
+
+	/*
+	 * PHY FW FSM EN and Direct FW write mechanism enable
+	 * FW FSM assumes that the data is present in the SRAM
+	 * The FSM will wait for SRAM init done before updating
+	 * the SRAM contents
+	 */
+	regmap_writel(9, mmPCIE_WRAP_PHY_FW_FSM);
+
+	/* Take the PHY out of reset */
+	regmap_writel(1, mmPSOC_GLOBAL_CONF_PCIE_PHY_RST_N);
+}
+
+void preboot_disable_pcie_ltr(void)
+{
+	u32 cap_hdr_cfg, dev_cap2_reg;
+
+	/* PCIe LTR configuration, set bits 31:20 to 0x1f8 */
+	cap_hdr_cfg = regmap_readl(mmPCIE_DBI_MARGIN_EXT_CAP_HDR_REG);
+	cap_hdr_cfg &= ~(PCIE_DBI_MARGIN_EXT_CAP_HDR_REG_NEXT_OFFSET_MASK);
+	cap_hdr_cfg |= (PCIE_CAP_HDR_NXT_OFFSET <<
+		 PCIE_DBI_MARGIN_EXT_CAP_HDR_REG_NEXT_OFFSET_SHIFT);
+	regmap_writel(cap_hdr_cfg, mmPCIE_DBI_MARGIN_EXT_CAP_HDR_REG);
+
+	/* Disable LTR support, set bit-11 to '0' */
+	dev_cap2_reg = regmap_readl(mmPCIE_DBI_DEVICE_CAPABILITIES2_REG);
+	dev_cap2_reg &= ~(PCIE_DBI_DEVICE_CAPABILITIES2_REG_PCIE_CAP_LTR_SUPP_MASK);
+	dev_cap2_reg |= (PCIE_LTR_DISABLE <<
+		 PCIE_DBI_DEVICE_CAPABILITIES2_REG_PCIE_CAP_LTR_SUPP_SHIFT);
+	regmap_writel(dev_cap2_reg, mmPCIE_DBI_DEVICE_CAPABILITIES2_REG);
+}
+
+u32 get_efuse_read_reg(void)
+{
+	return mmPSOC_EFUSE_READ_CTRL;
+}
+
+u32 preboot_device_id_get(void)
+{
+	/* TODO: remove hard coding */
+	return PCI_IDS_GAUDI2;
+}
+
+static void pcie_update_reg_range(u32 addr, u32 addr_end, u32 value)
+{
+	for (; addr <= addr_end ; addr += 4)
+		regmap_writel(value, addr);
+}
+
+static void pcie_wait_for_sram_init_done(void)
+{
+	uint8_t sram_init_done = 0;
+
+	do {
+		sram_init_done =
+			regmap_readl(mmPCIE_PHY_PHYX_SRAM_INIT_DONE_0);
+	} while (sram_init_done != 1);
+
+	sram_init_done = 0;
+	do {
+		sram_init_done =
+			regmap_readl(mmPCIE_PHY_PHYX_SRAM_INIT_DONE_1);
+	} while (sram_init_done != 1);
+
+	sram_init_done = 0;
+	do {
+		sram_init_done =
+			regmap_readl(mmPCIE_PHY_PHYX_SRAM_INIT_DONE_2);
+	} while (sram_init_done != 1);
+
+	sram_init_done = 0;
+	do {
+		sram_init_done =
+			regmap_readl(mmPCIE_PHY_PHYX_SRAM_INIT_DONE_3);
+	} while (sram_init_done != 1);
+}
+
+static void pcie_phy_set_sram_ext_ld_done(void)
+{
+	regmap_writel(1, mmPCIE_PHY_PHY_SRAM_EXT_LD_DONE);
+}
+
+static void pcie_phy_set_mpll_a_b_ssc_enable(void)
+{
+
+	pcie_update_reg_range(mmPCIE_PHY_PHYX_MPLLB_SSC_EN_0,
+		mmPCIE_PHY_PHYX_MPLLB_SSC_EN_3, 0x1);
+	pcie_update_reg_range(mmPCIE_PHY_PHYX_MPLLA_SSC_EN_0,
+		mmPCIE_PHY_PHYX_MPLLA_SSC_EN_3, 0x1);
+}
+
+static void pcie_phy_set_mpll_a_b_mul_bw(uint8_t separate_reference)
+{
+	uint32_t reg;
+
+	if (separate_reference) {
+		reg = 0xA0;
+		regmap_writel(0x00250025,
+			mmPCIE_PHY_PROTOCOL0_EXT_MPLLA_BANDWIDTH);
+		regmap_writel(0x00250025,
+			mmPCIE_PHY_PROTOCOL0_EXT_MPLLB_BANDWIDTH);
+	} else {
+		reg = 0x50;
+		regmap_writel(0x00530053,
+			mmPCIE_PHY_PROTOCOL0_EXT_MPLLA_BANDWIDTH);
+		regmap_writel(0x01C301C3,
+			mmPCIE_PHY_PROTOCOL0_EXT_MPLLB_BANDWIDTH);
+	}
+
+	regmap_writel(reg, mmPCIE_PHY_PROTOCOL0_EXT_MPLLA_MULTIPLIER);
+	regmap_writel(0x7D, mmPCIE_PHY_PROTOCOL0_EXT_MPLLB_MULTIPLIER);
+}
+
+static void pcie_phy_set_ext_tx_rx_ctrl(void) {}
+static void pcie_phy_set_protocol0_ext_tx_rx(void) {}
+
+static uint8_t get_srns_mode_from_strap(uint32_t boot_strap)
+{
+	uint32_t dump_sel = (boot_strap &
+		PSOC_GLOBAL_CONF_BOOT_STRAP_PINS_L_DUMP_SEL_MASK) >>
+		PSOC_GLOBAL_CONF_BOOT_STRAP_PINS_L_DUMP_SEL_SHIFT;
+
+	/* dumpsel[0] is SRNS */
+	return (dump_sel & DUMP_SEL_SRNS_MASK);
+}
+
+static uint8_t get_ssc_en_from_strap(uint32_t boot_strap)
+{
+	uint32_t dump_sel = (boot_strap &
+		PSOC_GLOBAL_CONF_BOOT_STRAP_PINS_L_DUMP_SEL_MASK) >>
+		PSOC_GLOBAL_CONF_BOOT_STRAP_PINS_L_DUMP_SEL_SHIFT;
+
+	/* dumpsel[1] is SSC EN */
+	return !!(dump_sel & DUMP_SEL_SSC_MASK);
+}
+
+uint8_t get_sris_mode_from_strap(void)
+{
+	uint32_t boot_strap;
+
+	boot_strap = regmap_readl(mmPSOC_GLOBAL_CONF_BOOT_STRAP_PINS_H);
+
+	return !!(boot_strap & PSOC_GLOBAL_CONF_BOOT_STRAP_PINS_H_SRIS_MODE_MASK);
+}
+
+static void pcie_wa_snps_case_01038891(void)
+{
+	uint8_t separate_reference = 0;
+	uint8_t ssc_en = 0, srns = 0;
+	uint32_t boot_strap = regmap_readl(mmPSOC_GLOBAL_CONF_BOOT_STRAP_PINS_L);
+
+	separate_reference = get_sris_mode_from_strap();
+	if (separate_reference) {
+		regmap_writel(1, mmPCIE_PHY_PIPE_RX_SRIS_MODE_EN);
+		regmap_writel(1, mmPCIE_AUX_APP_SRIS_MODE);
+	}
+	/* SRNS mode enable */
+	srns = get_srns_mode_from_strap(boot_strap);
+
+	if (!srns)
+		separate_reference = 1;
+
+	/* spread spectum clock enable */
+	ssc_en = get_ssc_en_from_strap(boot_strap);
+
+	/* TODO: Check fuse to see if SNPS WA needs to be enabled  */
+	if (ssc_en) {
+		pcie_phy_set_mpll_a_b_ssc_enable();
+
+		regmap_writel(0x5000, mmPCIE_PHY_EXT_MPLLB_SSC_PEAK);
+		regmap_writel(0x19CE,
+				mmPCIE_PHY_EXT_MPLLB_SSC_STEP_SIZE);
+		regmap_writel(0x6666, mmPCIE_PHY_EXT_MPLLA_SSC_PEAK);
+		regmap_writel(0x2108,
+				mmPCIE_PHY_EXT_MPLLA_SSC_STEP_SIZE);
+	}
+
+	pcie_phy_set_ext_tx_rx_ctrl();
+	pcie_phy_set_mpll_a_b_mul_bw(separate_reference);
+
+	pcie_phy_set_protocol0_ext_tx_rx();
+
+	regmap_writel(1, mmPCIE_PHY_PHY_EXT_CTRL_SEL);
+}
+
+__weak void preboot_init_pcie_phy(void)
+{
+	/* Reset PCIe PHY */
+	regmap_writel(0, mmPSOC_GLOBAL_CONF_PCIE_PHY_RST_N);
+
+	/*
+	 * Set PCIe PHY CR Para I/F to CFG.
+	 * Select CR PARA I/F instead of JTAG
+	 */
+	regmap_writel(1, mmPCIE_PHY_CR_PARA_SEL_0);
+	regmap_writel(1, mmPCIE_PHY_CR_PARA_SEL_1);
+	regmap_writel(1, mmPCIE_PHY_CR_PARA_SEL_2);
+	regmap_writel(1, mmPCIE_PHY_CR_PARA_SEL_3);
+
+	/* Execute synopsys workaround (01038891) */
+	pcie_wa_snps_case_01038891();
+
+	/* Select indirect access to CR PARA I/F instead of PRBI */
+	regmap_writel(1, mmPCIE_PHY_CR_PARA_CFG);
+
+	preboot_init_pcie_phy_fw_update();
+}
+
+static void preboot_pcie_device_id_vendor_id_set(void)
+{
+	u32 device_id_vendor_id = hl_pci_get_ven_id_dev_id();
+
+	/* Update Vendor ID */
+	regmap_writel(device_id_vendor_id,
+		      mmPCIE_DBI_DEVICE_ID_VENDOR_ID_REG);
+
+	/* Update subsystem Vendor ID */
+	regmap_writel(device_id_vendor_id,
+		      mmPCIE_DBI_SUBSYSTEM_ID_SUBSYSTEM_VENDOR_ID_REG);
+
+}
+
+void preboot_init_pcie(void)
+{
+	preboot_init_pcie_phy();
+
+	preboot_pcie_device_id_vendor_id_set();
+
+	/* Disable the EQ REDO */
+	regmap_writel(0x00402801, mmPCIE_DBI_GEN3_RELATED_OFF);
+
+	/* Wait for init_done before executing W/A */
+	pcie_wait_for_sram_init_done();
+
+	pcie_phy_set_sram_ext_ld_done();
+
+	/* Execute the Synopsys W/A */
+	regmap_writel(0xF, mmPCIE_PHY_CR_PARA_CTRL);
+	regmap_writel(0x0004A07D, mmPCIE_PHY_CR_PARA_CMD);
+	udelay(20);
+
+	/* PCIe disable LTR */
+	preboot_disable_pcie_ltr();
+
+	/* Enable and Mask LTSSM (erratum 102) */
+	regmap_writel(3, mmPCIE_AUX_LTSSM_EN);
+}
+void preboot_pcie_wait_for_link(void)
+{
+	u8 retry = HL_PCIE_LINK_RETRIES;
+
+	/* init dev id as soon as possible, while waiting for prstn */
+	preboot_device_id_init();
+
+	/* Wait for PRSTn */
+	while (!regmap_readl(mmPSOC_GLOBAL_CONF_PRSTN))
+		;
+
+	/* Only after PRSTn is de-asserted */
+	preboot_pcie_enable();
+
+	preboot_puts("Waiting for PCIe UP status\n");
+
+	while (1) {
+		unsigned int prstn, ltssm, smlh, rdlh;
+
+		prstn = regmap_readl(mmPSOC_GLOBAL_CONF_PRSTN);
+		ltssm = regmap_readl(mmPCIE_AUX_LTSSM_EN);
+		smlh = regmap_readl(mmPCIE_AUX_SMLH_LINK_UP);
+		rdlh = regmap_readl(mmPCIE_AUX_RDLH_LINK_UP);
+
+		if (prstn && smlh && rdlh) {
+			preboot_puts("PCIe link is UP!\n");
+			break;
+		}
+
+		/*
+		 * Do not wait indefinitely on PCIe link up status
+		 * Wait for 5 seconds (50 retries with 100 milli-
+		 * seconds sleep), else bailout.
+		 */
+		retry--;
+		if (!retry) {
+			preboot_puts("Timed-out waiting for PCIe link UP!\n");
+			break;
+		}
+
+		udelay(100000);
+
+		/*
+		 * Some host assert prstn during link up while moving to GEN3
+		 * This cause BAR size to be reset
+		 * We stop link up process by disabling ltssm
+		 * till BAR is reconfigured
+		 */
+		if (!prstn && ltssm) {
+			regmap_writel(0, mmPCIE_AUX_LTSSM_EN);
+			continue;
+		}
+
+		if (prstn && !ltssm) {
+			regmap_writel(3, mmPCIE_AUX_LTSSM_EN);
+			continue;
+		}
+	}
+
+	/* W/A H/W bug that may lead to have a wrong mask (0x1) in LTSSM */
+	regmap_writel(3, mmPCIE_AUX_LTSSM_EN);
+}
+
+void preboot_delay(void) {}
+
+unsigned int get_cs_pad_sel(void)
+{
+	return mmPSOC_GLOBAL_CONF_PAD_SEL_19;
+}
+
+void spi_ssi_cs0_get(unsigned int *mask, unsigned int *data_offset,
+		     unsigned int *out_en_offset)
+{
+	*mask = BIT(19 % 16);
+	*data_offset = mmPSOC_GPIO1_DATAOUT;
+	*out_en_offset = mmPSOC_GPIO1_OUTENSET;
+}
+
+void spi_base_get(unsigned int *regs)
+{
+	*regs = mmPSOC_QSPI_CTRLR0;
+}
+
+u32 hl_pll_timestamp_base_get(void)
+{
+	return mmCPU_TIMESTAMP_CNTCR;
+}
+
+void pci_dev_specific_hot_rst(void)
+{
+	regmap_writel(0x10, mmPCIE_CORE_HOT_RST);
+	regmap_writel(3, mmPCIE_AUX_LTSSM_EN);
+}
+
+void pci_dev_specific_flr(void)
+{
+	regmap_writel(1, mmPCIE_AUX_PF_FLR_ACTIVE);
+	regmap_writel(1, mmPCIE_AUX_PF_FLR_DONE);
+	regmap_writel(1, mmPCIE_AUX_FLR_INT);
+
+	/*
+	 * Clear PCIe SPI INTR. This is required to clear
+	 * FLR interrupt as recommended by HW team.
+	 */
+	regmap_writel(CPU_IF_PCIE_SPI_INTR_CLR_VAL_MASK,
+			mmCPU_IF_PCIE_SPI_INTR_CLR);
+}
